pkg/store/cache.go

```diff
@@ -115,10 +121,16 @@ func newIndexCache(reg prometheus.Registerer, maxBytes uint64) (*indexCache, err
 	return c, nil
 }
 
-func (c *indexCache) ensureFits(b []byte) {
+// ensureFits tries to make sure that the passed slice will fit into the LRU cache.
+// Returns true if it will fit.
+func (c *indexCache) ensureFits(b []byte) bool {
+	if uint64(len(b)) > c.maxSize {
+		return false
+	}
 	for c.curSize+uint64(len(b)) > c.maxSize {
 		c.lru.RemoveOldest()
```

Comment:
I think we could do more aggressive removal if the slice is big enough.

Reply:
Yes but how the implementation would be like? I have two ideas:
* If the slice size exceeds a certain threshold then we could remove 2 or more items at a time
* Get all keys, calculate their sizes, and remove the top X items to make sure there's enough size for the item `b`?

Reply:
Either way, maybe lets leave it for another day? I am mainly interested now in making sure that the loop doesn't go forever

Reply:
agree